### PR TITLE
Defer resolution of true and false branches in conditional types

### LINKED ISSUE #31302: Slow completionEntryDetails result in simple TS project
From https://github.com/microsoft/vscode/issues/73311

**TypeScript Version:** 3.5.0-dev.20190507

**Search Terms:**
- suggestion
- completionEntryDetails
- tsserver

**Repo**

```ts
git clone https://github.com/mjbvz/vscode-73311
cd vscode-73311
npm install
code-insiders .
```

1. Open `index.js` and try triggering suggestions after `Flex` by deleting the trailing `x` then typing it again

The project uses styled components:

```ts
import styled from "styled-components";

type Justify = "space-around" | "space-between";

export const FlexBox = styled.div<{
    column?: boolean;
    justify?: Justify;
}>`
  display: flex;
  flex-direction: ${props => (props.column ? "column" : "row")};
  ${props => !!props.justify && `justify-content: ${props.justify}`}
`;

export const Flex = styled.div<{ flex?: number }>`
  flex: ${props => props.flex || 1};
`; 
```

**Bug**
The `completionEntryDetails` request this triggers can often take half a second or longer to process:

```
Info 214  [15:56:42.198] request:
    {"seq":37,"type":"request","command":"completionEntryDetails","arguments":{"file":"/Users/matb/projects/san/index.ts","line":1,"offset":5,"entryNames":[{"name":"Flex","source":"/Users/matb/projects/san/components"}]}}
Info 215  [15:56:42.199] getCompletionData: Get current token: 0
Info 216  [15:56:42.199] getCompletionData: Is inside comment: 0
Info 217  [15:56:42.199] getCompletionData: Get previous token 1: 0
Info 218  [15:56:42.199] getCompletionData: Get previous token 2: 0
Info 219  [15:56:42.202] getCompletionData: Semantic work: 3
Perf 220  [15:56:43.103] 37::completionEntryDetails: elapsed time (in milliseconds) 905.5459
Info 221  [15:56:43.105] response:
    {"seq":0,"type":"response","command":"completionEntryDetails","request_seq":37,"success":true,"body":[{"name":"Flex","kindModifiers":"export","kind":"const","displayParts":[{"text":"const","kind":"keyword"},{"text":" ","kind":"space"},{"text":"Flex","kind":"localName"},{"text":":","kind":"punctuation"},{"text":" ","kind":"space"},{"text":"StyledComponent","kind":"aliasName"},{"text":"<","kind":"punctuation"},{"text":"\"div\"","kind":"stringLiteral"},{"text":",","kind":"punctuation"},{"text":" ","kind":"space"},{"text":"any","kind":"keyword"},{"text":",","kind":"punctuation"},{"text":" ","kind":"space"},{"text":"{","kind":"punctuation"},{"text":"\n","kind":"lineBreak"},{"text":"    ","kind":"space"},{"text":"flex","kind":"propertyName"},{"text":"?","kind":"punctuation"},{"text":":","kind":"punctuation"},{"text":" ","kind":"space"},{"text":"number","kind":"keyword"},{"text":";","kind":"punctuation"},{"text":"\n","kind":"lineBreak"},{"text":"}","kind":"punctuation"},{"text":",","kind":"punctuation"},{"text":" ","kind":"space"},{"text":"never","kind":"keyword"},{"text":">","kind":"punctuation"}],"documentation":[],"codeActions":[{"description":"Import 'Flex' from module \"./components\"","changes":[{"fileName":"/Users/matb/projects/san/index.ts","textChanges":[{"start":{"line":1,"offset":1},"end":{"line":1,"offset":1},"newText":"import { Flex } from \"./components\";\n\n"}]}]}],"source":[{"text":"./components","kind":"text"}]}]}
```

**Playground Link:** 

**Related Issues:**

### LINKED ISSUE #31341: Newest version 1.33.1 has electron_node tsserver.js eating up 100% cpu randomly
So this seems to be a recurring issue over the years. But there also appears to be no definitive solution either.

Here's what my process manager is showing:
Screen Shot 2019-04-19 at 11 14 04 AM

Not entirely sure why. I'm not coding in Typescript... I'm not using any clear extension that's typescript related... unless ESLint does this, which I wouldnt be surprised about.

I have reverted back to version 1.32 and everything runs like a dream. 

Intellisense is FAST.
No code highlighting issues.

I just wanted to bring this back up. 

I'm working on a tight deadline for two sites so once I get these done I will try to get my tslog info... but in the meantime I wanted to see if anybody else has experienced this and found a solution that I have not come across yet. 

Thanks!

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR restores our previous behavior of deferring resolution of the true and false branches of conditional types. This fixes the performance regression we were seeing with statement completion in projects using `styled-components` (performance is now similar to 3.3).

Because resolution of the true and false branches is deferred, simplifications of conditional types of the form `T extends xxx ? T : never` and `T extends xxx ? never : T` only occur when the actual declarations of the conditional types are written with identical types in the `T` position and type `never` in the other position (i.e. simplification doesn't occur when conditional types only match the pattern following instantiation). This shouldn't matter in practice as the types we really care about, `Exclude` and `Extract`, follow the pattern. For example, #28748 is still fixed.

Fixes #31302.
Fixes #31341.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot test this

**Comment by ahejlsberg** [comment]
@typescript-bot run dt

**Comment by ahejlsberg** [comment]
There are definitely significantly fewer types being created when the true and false branches are deferred.

Compiling `styled-components` with current master:

```
Files:            28
Lines:         21723
Nodes:         60563
Identifiers:   19996
Symbols:      129415
Types:        132769
Memory used: 207569K
I/O read:      0.01s
I/O write:     0.00s
Parse time:    0.40s
Bind time:     0.17s
Check time:    8.36s
Emit time:     0.00s
Total time:    8.92s
```

Compiling `styled-components` with this PR:

```
Files:            28
Lines:         21723
Nodes:         60563
Identifiers:   19996
Symbols:      121459
Types:         87840
Memory used: 168256K
I/O read:      0.01s
I/O write:     0.00s
Parse time:    0.39s
Bind time:     0.17s
Check time:    6.21s
Emit time:     0.00s
Total time:    6.77s
```

Compiling `styled-components` with 3.3.4:

```
Files:            28
Lines:         21776
Nodes:         61201
Identifiers:   20271
Symbols:      102024
Types:         63544
Memory used: 120873K
I/O read:      0.01s
I/O write:     0.00s
Parse time:    0.36s
Bind time:     0.18s
Check time:    4.84s
Emit time:     0.00s
Total time:    5.39s
```

**Comment by j-oliveras** [comment]
Probably fixes #31341 too.

**Comment by weswigham** [comment]
@ahejlsberg we should be able to defer the simplification to assignability checking, similar to indexed accesses, to keep the fixed use-cases functional.

**Comment by weswigham** [comment]
@ahejlsberg #31374 is open with what I suggested - perf impact (compared to this PR as-is) in `styled-components` is either minimal or positive, and all the simplifications continue to work as they do in `master`.

**Comment by ahejlsberg** [comment]
@typescript-bot test this

**Comment by ahejlsberg** [comment]
@typescript-bot run dt

File: tests/baselines/reference/conditionalTypes2.errors.txt [hunk 1]
```diff
@@ -10,13 +10,13 @@ tests/cases/conformance/types/conditional/conditionalTypes2.ts(24,5): error TS23
...
-                  Type 'string' is not assignable to type 'number | "toString" | "charAt" | "charCodeAt" | "concat" | "indexOf" | "lastIndexOf" | "localeCompare" | "match" | "replace" | "search" | "slice" | "split" | "substring" | "toLowerCase" | "toLocaleLowerCase" | "toUpperCase" | "toLocaleUpperCase" | "trim" | "length" | "substr" | "valueOf"'.
-                    Type 'keyof B' is not assignable to type '"valueOf"'.
-                      Type 'string | number | symbol' is not assignable to type '"valueOf"'.
-                        Type 'string' is not assignable to type '"valueOf"'.
+            Type 'string' is not assignable to type 'number | "toString" | "charAt" | "charCodeAt" | "concat" | "indexOf" | "lastIndexOf" | "localeCompare" | "match" | "replace" | "search" | "slice" | "split" | "substring" | "toLowerCase" | "toLocaleLowerCase" | "toUpperCase" | "toLocaleUpperCase" | "trim" | "length" | "substr" | "valueOf" | keyof A'.
```
Comment (weswigham) [inline tests/baselines/reference/conditionalTypes2.errors.txt]: did `keyof A` expand into... itself o.O

Comment (ahejlsberg) [reply tests/baselines/reference/conditionalTypes2.errors.txt]: In the first `keyof A`, `A` is a substitution type. It then gets expanded into `keyof (A & string)` which becomes the union that ends in `keyof A` where `A` is the actual type variable. Not ideal, but neither was the prior error message.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -10258,6 +10260,38 @@ namespace ts {
...
+            const extendsType = type.extendsType;
+            const trueType = getTrueTypeFromConditionalType(type);
+            const falseType = getFalseTypeFromConditionalType(type);
+            // Simplifications for types of the form `T extends U ? T : never` and `T extends U ? never : T`.
+            if (falseType.flags & TypeFlags.Never && getActualTypeVariable(trueType) === getActualTypeVariable(checkType)) {
+                if (checkType.flags & TypeFlags.Any || isTypeAssignableTo(getRestrictiveInstantiation(checkType), getRestrictiveInstantiation(extendsType))) { // Always true
+                    return getSimplifiedType(trueType, writing);
+                }
+                else if (isIntersectionEmpty(checkType, extendsType)) { // Always false
+                    return neverType;
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Why wasn't this just a `Debug.fail`?

Comment (ahejlsberg) [reply src/compiler/checker.ts]: Because it can happen and is perfectly fine when it does. Or am I misunderstanding your question?

Comment (ahejlsberg) [reply src/compiler/checker.ts]: BTW, the comment "Always false" here refers to when the condition of the conditional type is known to always be false. It doesn't imply that the `if` condition is always false.

**Pull request merged**